### PR TITLE
Check variable debt token

### PR DESCRIPTION
**Pull request by gadicer**
Add some rules and a helper invariant to cover a bug injected to _accrueDebtOnAction()

### CODE REVIEW

File: certora/scripts/verifyGhoDiscountRateStrategy.sh
Comment (gadicer) [inline certora/scripts/verifyGhoDiscountRateStrategy.sh:1]: certora/scripts/verifyGhoVariableDebtToken.sh remains the same as in branch cvl2

File: certora/harness/ghoVariableDebtTokenHarness_internal.sol [hunk 1]
```diff
@@ -0,0 +1,57 @@
+pragma solidity 0.8.10;
+
+import {GhoVariableDebtToken} from '../munged/contracts/facilitators/aave/tokens/GhoVariableDebtToken.sol';
+import {WadRayMath} from '@aave/core-v3/contracts/protocol/libraries/math/WadRayMath.sol';
+import {IPool} from '@aave/core-v3/contracts/interfaces/IPool.sol';
+
+contract GhoVariableDebtTokenHarness is GhoVariableDebtToken {
```
Comment (MichaelMorami) [inline certora/harness/ghoVariableDebtTokenHarness_internal.sol]: the contract need to be called harnessInternal like the file.
It should inherit from harness.

File: certora/specs/ghoVariableDebtToken_internal.spec [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	balanceIncrease, discountScaled = accrueDebtOnAction(e, user,previousScaledBalance,discountPercent,index);
+	 uint256 accumulated_interest_after = getUserAccumulatedDebtInterest(user);
+	
+	assert 	ray() <= user_index_before
+			&& to_mathint(user_index_before + ray()) < to_mathint(index) // user index increase by more than 1 ray
+			&& 0 < previousScaledBalance
+			&& discountPercent < 5000 // discount rate is less than 50% 
+					=> balanceIncrease > 0;
+
+	assert 	balanceIncrease > 0 => accumulated_interest_after > accumulated_interest_before;
```
Comment (MichaelMorami) [inline certora/specs/ghoVariableDebtToken_internal.spec]: what about an assert that additional data is equal current index?

File: certora/specs/ghoVariableDebtToken_internal.spec [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+	env e;
+	address user;
+    uint256 previousScaledBalance; uint256 discountPercent; uint256 index;
+	uint256 balanceIncrease; uint256 discountScaled;
+	uint256 user_index_before = getUserCurrentIndex(user);
+	uint256 accumulated_interest_before = getUserAccumulatedDebtInterest(user);
+	balanceIncrease, discountScaled = accrueDebtOnAction(e, user,previousScaledBalance,discountPercent,index);
+	 uint256 accumulated_interest_after = getUserAccumulatedDebtInterest(user);
+	
+	assert 	ray() <= user_index_before
```
Comment (MichaelMorami) [inline certora/specs/ghoVariableDebtToken_internal.spec]: This should be changed imo

File: certora/specs/ghoVariableDebtToken.spec [hunk 4]
```diff
@@ -129,6 +130,76 @@ invariant discountCantExceed100Percent(address user)
...
+	env e;
+	address newDiscountToken; 
+  	updateDiscountToken(e, newDiscountToken);
+	assert newDiscountToken != 0;
+}
+
+
+
+// check user index after mint()
+rule user_index_after_mint
```
Comment (MichaelMorami) [inline certora/specs/ghoVariableDebtToken.spec:164]: use filter with all function that call accrue

**Pull request merged**